### DISCUSSION TITLE
Structured logging and various logging improvements

### DISCUSSION THREAD

**Discussion by raskyld** [seaweedfs/seaweedfs#6087] [Ideas]
Hello!

We've started experimenting with SeaweedFS a bit and I was wondering if you would accept a PR to setup structured logging in the various components using the Standard Library: https://pkg.go.dev/log/slog (or another if you will).

We've been running without replication and with a single volume without any problems, but we faced various challenges going with replication and 3 volumes instead. Debugging them was quite painful since the logs didn't help much (e.g. to understand why a volume has become read-only), even with increased verbosity, I think some debug logs could be added in different part of the codebase.

Before I start working on that, are you willing to review a PR that would bring those changes in the codebase? 🙏 

Thanks a lot for your work! 💪

**Comment by chrislusf** [comment]
slog is a politically correct choice, but I found it is lacking in these aspects (please correct me if wrong):
* there are no file name and line number by default.
* there are computational cost to evaluate the log parameters.

Also, current logs are formatted. Changing to structured logs would need a lot of engineering work. And hard to tell whether there are bugs introduced for the transition.

**Reply by raskyld** [reply]
>there are no file name and line number by default.

Actually, the PC (program counter) is stored in every record, please, see: https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/log/slog/record.go;l=36

So I am *pretty* sure we can ask the default Handler to print it with every records.

> there are computational cost to evaluate the log parameters.

I am not an expert of the logging libraries landscape so I cannot talk too much about the performance costs, I know it uses lazy evaluation (i.e. print only if log level enables it).

I know a part of the community is not satisfied with `slog` and tbh, I am not opinionated about which exact library to use. 
(I know a pretty good one that is used in production by "big ones": https://github.com/rs/zerolog).

> Also, current logs are formatted. 

Do you mean there are implementations that depends on the current format to parse logs?

---

My main motivations were to make it easier to do some analysis on the logs, that includes:

- Outputting logs that can be parsed by main-stream log ingestion platform so you can do visualisation later. (e.g. I want to have a dashboard where I can see all volume transition from writable/readonly, it's easier to do if I have some fields like `volumeId=XX` and `state=writable/readonly`)
- Adding some record and context in places to help with debugging wherever possible (e.g. using the previous example, adding a `reason` enum to explain why the volume changed to readonly, e.g. `NotEnoughReplicas`)
- Make users choose between readability and performance trade-off when they start `weed`. For example, in `zerolog`, the default format is optimized for performance but you can enable a human-format if wanted: https://github.com/rs/zerolog?tab=readme-ov-file#pretty-logging.

**Reply by chrislusf** [reply]
For example, this line https://github.com/seaweedfs/seaweedfs/blob/master/weed/server/master_server.go#L183 uses formatted logs.

It is not straight forward to change to slog.

**Reply by raskyld** [reply]
Oh you meant, some "fields" need to be rendered in a specific format!
For that, the `slog` approach would be to wrap the values in a struct that implements https://pkg.go.dev/log/slog#LogValuer.

It would only be evaluated if the log level enable this line.

**Reply by chrislusf** [reply]
anyway, it would be a lot of code change.

**Reply by raskyld** [reply]
If you think it's not worth the benefits I listed, feel free to close so.
